src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_code.go

```diff
@@ -172,62 +175,56 @@ func (a *StoragePowerActorCode_I) GetTotalPower(rt Runtime) block.StoragePower {
 	return totalPower
 }
 
-func (a *StoragePowerActorCode_I) EnsurePledgeCollateralSatisfied(rt Runtime) bool {
-
-	ret := false
-
-	// TODO: convert msgSender to MinerActorID
-	var minerID addr.Address
+func (a *StoragePowerActorCode_I) EnsurePledgeCollateralSatisfied(rt Runtime) InvocOutput {
 
 	h, st := a.State(rt)
+	ret := st._ensurePledgeCollateralSatisfied(rt)
+	UpdateRelease(rt, h, st)
 
-	powerEntry, found := st.PowerTable()[minerID]
-
-	if !found {
-		rt.Abort("miner not found.")
-	}
-
-	pledgeCollateralRequired := st._getPledgeCollateralReq(rt, powerEntry.ActivePower()+powerEntry.InactivePower())
-
-	if pledgeCollateralRequired < powerEntry.LockedPledgeCollateral() {
-		ret = true
-	} else if pledgeCollateralRequired < (powerEntry.LockedPledgeCollateral() + powerEntry.AvailableBalance()) {
-		st._lockPledgeCollateral(rt, minerID, (pledgeCollateralRequired - powerEntry.LockedPledgeCollateral()))
-		ret = true
+	if !ret {
+		return rt.ErrorReturn(exitcode.InsufficientPledgeCollateral)
 	}
 
-	UpdateRelease(rt, h, st)
-
-	return ret
+	return rt.SuccessReturn()
 }
 
-func (a *StoragePowerActorCode_I) ProcessFaultReport(rt Runtime, report sector.FaultReport) {
+// slash pledge collateral for Declared, Detected and Terminated faults
+func (a *StoragePowerActorCode_I) SlashPledgeForStorageFault(rt Runtime, affectedPower block.StoragePower, faultType sector.StorageFaultType) {
 
 	var msgSender addr.Address // TODO replace this
 
-	h, st := a.State(rt)
+	// placeholder values
+	const unitDeclaredFaultSlash = 1    // placeholder
+	const unitDetectedFaultSlash = 5    // placeholder
+	const unitTerminatedFaultSlash = 10 // placeholder
 
-	declaredFaultSlash := report.GetDeclaredFaultSlash()
-	detectedFaultSlash := report.GetDetectedFaultSlash()
-	terminatedFaultSlash := report.GetTerminatedFaultSlash()
+	h, st := a.State(rt)
 
-	st._slashPledgeCollateral(rt, msgSender, (declaredFaultSlash + detectedFaultSlash + terminatedFaultSlash))
+	switch faultType {
+	case sector.DeclaredFault:
+		amountToSlash := actor.TokenAmount(unitDeclaredFaultSlash * affectedPower)
```

Comment:
These expressions all need to be rewritten as proportions of the miner's pledge collateral requirement, which is dynamic.